flag/pflag.go

```diff
@@ -0,0 +1,365 @@
+
+	s := PflagSet{
+		Flags:           pflag.CommandLine,
+		ParseFunc:       func() error { flag.Parse(); return nil },
+		ptrType:         ptyp,
+		flagsRegistered: true,
+		NameCfg:         cfg,
+		flagFieldName:   map[string]string{},
+		flagValues:      map[string]interface{}{},
+	}
+
+	if err := s.registerPFlags(pval, ptyp); err != nil {
+		return nil, err
+	}
+
+	return &s, nil
+}
+
+// NewPFlagSetWithArgs creates a new FlagSet and registers flags in it
+func NewPFlagSetWithArgs(cfg *NameConfig, template interface{}, args []string) (*PflagSet, error) {
+	pval, ptyp, ptrifyErr := ptrified(template)
+	if ptrifyErr != nil {
+		return nil, ptrifyErr
+	}
+
+	fs := pflag.NewFlagSet("", pflag.ContinueOnError)
+
+	s := PflagSet{
+		Flags:           fs,
+		ParseFunc:       func() error { return fs.Parse(args) },
+		ptrType:         ptyp,
+		flagsRegistered: true,
+		NameCfg:         cfg,
+		flagFieldName:   map[string]string{},
+		flagValues:      map[string]interface{}{},
+	}
+
+	if err := s.registerPFlags(pval, ptyp); err != nil {
+		return nil, err
+	}
+
+	return &s, nil
+}
+
+// PflagSet is a flagset. Please only use PflagSet when using cobra command line
+// tool. Use Set for flags functionality for other cases
+type PflagSet struct {
+	Flags     *pflag.FlagSet
+	ParseFunc func() error
+
+	ptrType reflect.Type
+
+	// NameCfg defines tunables for constructing flag-names
+	NameCfg *NameConfig
+
+	flagsRegistered bool
+	tfmr            *transform.Transformer
+	trnslVal        reflect.Value
+	// Map to store the flag name (key) and field name (value)
+	flagFieldName map[string]string
+	flagValues    map[string]interface{}
+}
+
+func (s *PflagSet) parse() error {
+	if s.ParseFunc == nil {
+		return fmt.Errorf("unparsed flagset with no ParseFunc set")
+	}
+	if err := s.ParseFunc(); err != nil {
+		return fmt.Errorf("failed to parse pflags: %s", err)
+	}
+	return nil
+}
+
+func (s *PflagSet) registerPFlags(tmpl reflect.Value, ptyp reflect.Type) error {
+	fm := transform.NewFlattenMangler(transform.DialsTagName, s.NameCfg.FieldNameEncodeCasing, s.NameCfg.TagEncodeCasing)
+	tfmr := transform.NewTransformer(ptyp, fm)
+	val, TrnslErr := tfmr.Translate()
+	if TrnslErr != nil {
+		return TrnslErr
+	}
+
+	s.tfmr = tfmr
+	s.trnslVal = val
+
+	t := val.Type()
+
+	k := t.Kind()
+	for k == reflect.Ptr {
+		t = t.Elem()
+		k = t.Kind()
+	}
+
+	// the input kind will be struct after calling Translate on it
+	for i := 0; i < t.NumField(); i++ {
+		sf := t.Field(i)
+		help := DefaultFlagHelpText
+		if x, ok := sf.Tag.Lookup(HelpTextTag); ok {
+			help = x
+		}
+
+		name := s.mkname(sf)
+		s.flagFieldName[name] = sf.Name
+
+		// if the flag already exists, don't register so the user can override
+		// our behavior
+		if s.Flags.Lookup(name) != nil {
+			continue
+		}
+
+		ft := sf.Type
+
+		k := ft.Kind()
+		for k == reflect.Ptr {
+			ft = ft.Elem()
+			k = ft.Kind()
+		}
+		isValue := ft.Implements(flagReflectType) || reflect.PtrTo(ft).Implements(flagReflectType)
```

Comment:
we should be checking whether the type implements [`pflag.Value`](https://pkg.go.dev/github.com/spf13/pflag?tab=doc#Value) here.

Reply:
Fixed